zcash_client_backend/src/data_api/wallet.rs

```diff
@@ -303,19 +303,27 @@ where
         )));
     }
 
+    let (_, change_address) = dfvk.change_address();
+
     // Create the transaction
     let mut builder = Builder::new(params.clone(), height);
     for selected in spendable_notes {
-        let from = {
+        // If the diversifier corresponds to that for the internal change address, use that
+        // address to construct the note, otherwise derive the external address corresponding
+        // to the diversifier for the selected note.
+        //
+        // FIXME: This only works so long as the wallet never chooses anything but the default
+        // diversifier index for creating a change address.
```

Comment:
I agree that this shouldn't occur in general (for wallets following ZIP 316), but I'd prefer if we could check internal vs external reliably. And I've thought of a hacky way to do it 🙂
```rust
let note = {
    let external_note = dfvk.fvk()
        .vk
        .to_payment_address(selected.diversifier)
        .unwrap()
        .create_note(selected.note_value.into(), selected.rseed);
    let internal_note = dfvk.internal()
        .fvk()
        .vk
        .to_payment_address(selected.diversifier)
        .unwrap()
        .create_note(selected.note_value.into(), selected.rseed);

    let expected_root = selected.witness.root();
    if expected_root == merkle_path.root(external_note.cmu()) {
        Ok(external_note)
    else if expected_root == merkle_path.root(internal_note.cmu()) {
        Ok(internal_note)
    } else {
        Err(TODO) 
    }
}?;
```